### PR TITLE
fix(ivy): sync view with blueprint when necessary

### PR DESCRIPTION
**Pull request by kara**
This fixes a bug where blueprints and view instances were getting out of sync.

### CODE REVIEW

File: packages/core/src/render3/instructions.ts [hunk 1]
```diff
@@ -2230,10 +2233,44 @@ export function componentRefresh<T>(adjustedElementIndex: number): void {
...
+ * 3. Second <comp> is matched as a component and its LViewData is created.
+ * 4. App template completes processing, so it's time to check child templates.
+ * 5. First <comp> template is checked. It has a directive, so its def is pushed to blueprint.
+ * 6. Second <comp> template is checked. Its blueprint has been updated by the first
+ * <comp> template, but its LViewData was created before this update, so it is out of sync.
+ *
+ * Note that embedded views inside ngFor loops will never be out of sync because these views
+ * are processed as soon as they are created.
+ *
+ * @param The view to sync
```
Comment (alfaproject) [inline packages/core/src/render3/instructions.ts]: Missing the param name: `componentView`

File: packages/core/src/render3/di.ts [hunk 2]
```diff
@@ -118,10 +119,11 @@ export function getOrCreateNodeInjectorForNode(
   const parentData = parentView[TVIEW].data as any;
   const injectorIndex = tNode.injectorIndex;
 
-  for (let i = 0; i < PARENT_INJECTOR; i++) {
-    const bloomIndex = parentIndex + i;
-    hostView[injectorIndex + i] =
-        parentLoc === -1 ? 0 : parentView[bloomIndex] | parentData[bloomIndex];
+  if (parentLoc !== -1) {
```
Comment (mhevery) [inline packages/core/src/render3/di.ts]: add a comment which explains what the magical `-1` means

File: packages/core/src/render3/di.ts [hunk 3]
```diff
@@ -118,10 +119,11 @@ export function getOrCreateNodeInjectorForNode(
...
+    for (let i = 0; i < PARENT_INJECTOR; i++) {
+      const bloomIndex = parentIndex + i;
```
Comment (mhevery) [inline packages/core/src/render3/di.ts]: I think this merges the parent bloom with the cumulative bloom. Please add a comment to that effect.

File: packages/core/src/render3/di.ts [hunk 4]
```diff
@@ -106,8 +106,9 @@ export function getOrCreateNodeInjectorForNode(
   if (tView.firstTemplatePass) {
     // TODO(kara): Store node injector with host bindings for that node (see VIEW_DATA.md)
     tNode.injectorIndex = hostView.length;
-    tView.blueprint.push(0, 0, 0, 0, 0, 0, 0, 0, null);  // foundation for cumulative bloom
-    tView.data.push(0, 0, 0, 0, 0, 0, 0, 0, tNode);      // foundation for node bloom
+    tView.data.push(0, 0, 0, 0, 0, 0, 0, 0, tNode);  // foundation for node bloom
```
Comment (mhevery) [inline packages/core/src/render3/di.ts]: refactor to 
```
function setupBloom(array, node) {
  array.push(0, 0, 0, 0, 0, 0, 0, 0, node);
}

setupBloom(tView.data, tNode); 
setupBloom(hostView, null);
setupBloom(tView.blueprint, null);
```

**Review by mhevery — APPROVED** [review]
With comments

**Review by mhevery — APPROVED** [review]

**Comment by kara** [comment]
[presubmit](http://test/OCL:215942293:BASE:215942350:1538765722690:d22048b0)

**Pull request closed without merging**